### PR TITLE
[BC Break] Make DomainCollectionInterface::map() returns an instance of self

### PR DESCRIPTION
**Pull request by Lctrs**
Also make this method lazy in DomainCollection

### CODE REVIEW

**Comment by ro0NL** [comment]
looks reasonable :) i got a bit inspired by java: https://stackoverflow.com/questions/2651819/why-doesnt-java-map-extends-collection

But in PHP we're finy anyway.. ill give this some thought :+1:

**Comment by Lctrs** [comment]
Main goal of this PR was to make `DomainCollection::map()` lazy. If you don't want to make DomainCollectionInterface::map()`` returning an instance of self, we can always turn its return into `iterable`.

**Comment by ro0NL** [comment]
iterable still implies a list/collection sort of. To avoid that and have laziness it could be `Collection::map(): ArrayAccess` as well :thinking: 

Basically i dont want to sell "can be foreached" for a map

**Comment by ro0NL** [comment]
:+1: if you can do something like

`return new class implements ArrayAccess {}`

**Review by ro0NL — APPROVED** [review]
actually i think it's fine as is :sweat: we're just mapping the values from a known collection; not transforming a list into a map sort of speak.

Can you update the references in docs/ as well?

File: src/User/Role/UserRoleProvider.php [hunk 1]
```diff
@@ -25,6 +25,6 @@ public function getRoles(User $user): array
             return $userRole->getRoleName();
         });
 
-        return array_values($roles);
+        return array_values(iterator_to_array($roles));
```
Comment (ro0NL) [inline src/User/Role/UserRoleProvider.php]: can be `return iterator_to_array($roles, false)` now :)

**Comment by Lctrs** [comment]
@ro0NL Done.

**Review by ro0NL — APPROVED** [review]

**Comment by ro0NL** [comment]
Thanks @Lctrs

**Pull request merged**